lightning/src/ln/channelmanager.rs

```diff
@@ -5896,66 +5895,41 @@ where
-												let purpose = events::PaymentPurpose::SpontaneousPayment(preimage);
-												check_total_value!(purpose);
+								let payment_preimage = if has_recipient_created_payment_secret {
+									if let Some(ref payment_data) = payment_data {
+										let (payment_preimage, min_final_cltv_expiry_delta) = match inbound_payment::verify(payment_hash, &payment_data, self.highest_seen_timestamp.load(Ordering::Acquire) as u64, &self.inbound_payment_key, &self.logger) {
+											Ok(result) => result,
+											Err(()) => {
+												log_trace!(self.logger, "Failing new HTLC with payment_hash {} as payment verification failed", &payment_hash);
+												fail_htlc!(claimable_htlc, payment_hash);
 											}
-										}
-									},
-									hash_map::Entry::Occupied(inbound_payment) => {
-										if let OnionPayload::Spontaneous(_) = claimable_htlc.onion_payload {
-											log_trace!(self.logger, "Failing new keysend HTLC with payment_hash {} because we already have an inbound payment with the same payment hash", &payment_hash);
-											fail_htlc!(claimable_htlc, payment_hash);
-										}
-										let payment_data = payment_data.unwrap();
-										if inbound_payment.get().payment_secret != payment_data.payment_secret {
-											log_trace!(self.logger, "Failing new HTLC with payment_hash {} as it didn't match our expected payment secret.", &payment_hash);
-											fail_htlc!(claimable_htlc, payment_hash);
-										} else if inbound_payment.get().min_value_msat.is_some() && payment_data.total_msat < inbound_payment.get().min_value_msat.unwrap() {
-											log_trace!(self.logger, "Failing new HTLC with payment_hash {} as it didn't match our minimum value (had {}, needed {}).",
-												&payment_hash, payment_data.total_msat, inbound_payment.get().min_value_msat.unwrap());
-											fail_htlc!(claimable_htlc, payment_hash);
-										} else {
-											let purpose = events::PaymentPurpose::from_parts(
-												inbound_payment.get().payment_preimage,
-												payment_data.payment_secret,
-												payment_context,
-											);
-											let payment_claimable_generated = check_total_value!(purpose);
-											if payment_claimable_generated {
-												inbound_payment.remove_entry();
+										};
+										if let Some(min_final_cltv_expiry_delta) = min_final_cltv_expiry_delta {
+											let expected_min_expiry_height = (self.current_best_block().height + min_final_cltv_expiry_delta as u32) as u64;
+											if (cltv_expiry as u64) < expected_min_expiry_height {
+												log_trace!(self.logger, "Failing new HTLC with payment_hash {} as its CLTV expiry was too soon (had {}, earliest expected {})",
+												&payment_hash, cltv_expiry, expected_min_expiry_height);
+												fail_htlc!(claimable_htlc, payment_hash);
 											}
 										}
+										payment_preimage
+									} else { debug_assert!(false); None }
```

Comment:
```suggestion
									} else { debug_assert!(false); None }
```
why we do this dabug assert here? and not just panic always or emit a log if we are not panicking in production? I feel this can be hard to track down if this happen in some wallet in production, is it?

Reply:
We tend to add `debug_assert`s for cases that should be unreachable, if that makes sense. So we should never hit this.